### ISSUE TITLE
Python Bug: lambda function refers only one environment

### ISSUE THREAD

**Issue by maguro27** [aai-institute/tianshou#1155]
- [ ] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [x] documentation request (i.e. "X is missing from the documentation.")
    + [ ] new feature request
    + [x] design request (i.e. "X should be changed to Y.")
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, gymnasium as gym, torch, numpy, sys
  print(tianshou.__version__, gym.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  0.5.1 0.29.1 2.3.0a0+40ec155e58.nv24.03 1.24.4 3.10.12 (main, Nov 20 2023, 15:14:05) [GCC 11.4.0] linux
  ```

The tutorial uses the lambda function for making callable environment functions many times.
However, I confirmed the Python bug (Python 3.10.12) when I use user-defined environment that is shown as follows,
```python
train_envs: list[gym.Env] = making_my_env() # e.g., len(train_envs) == 4
ts_train_envs = DummyVectorEnv([lambda: env for env in train_envs])
tmp_envs = [lambda: env for env in train_envs]

for env in tmp_envs:
    print(env().reset())
for env in ts_train_envs._env_fns:
    print(env().reset())
```
Then, I can get the same return values of the environment, but stacked 4.
Hence, I fix the above code as follows,
```python
def callable_env(env: gym.Env) -> Callable:
    def _callable_env() -> gym.Env:
        return env

    return _callable_env

train_envs: list[gym.Env] = making_my_env() # e.g., len(train_envs) == 4
ts_train_envs = DummyVectorEnv([callable_env(env) for env in train_envs])
tmp_envs = [callable_env(env) for env in train_envs]

for env in tmp_envs:
    print(env().reset())
for env in ts_train_envs._env_fns:
    print(env().reset())
```
This works properly.
In conclusion, I suggest that the tutorial should not use the lambda function.

**Comment by MischaPanch** [comment]
You are using a very old version of tianshou. Could you pls try on either 1.0.0 or on the version on master?

**Comment by maguro27** [comment]
@MischaPanch 
I update the python and tianshou version.
>1.0.0 0.28.1 2.3.0+cu121 1.24.4 3.11.9 (main, Apr  6 2024, 17:59:24) [GCC 9.4.0] linux

However, this lambda function issue remains.

**Comment by dantp-ai** [comment]
Hi @maguro27,

It seems that `env` is not looked up until the lambda function is called, but by the end of the loop, `env` is bound to the last element in the list, hence you get the last environment four times. You can read more about it and closures with lambdas [here](https://stackoverflow.com/questions/28268439/python-list-comprehension-with-lambdas).

This should now work as expected since the default value for `env` is evaluated when the lambda function is defined:
```
ts_train_envs = DummyVectorEnv([lambda: env=env for env in train_envs])
```

Which Tianshou tutorial are you looking at?

**Comment by maguro27** [comment]
@dantp-ai 

Thank you for your comments.
I understand I misunderstood the behavior of the lambda function.

Tianshou tutorials only use default gym environments.
Hence, I misunderstand the behavior.
Therefore, I think maintainers might want to add information for using custom environments (e.g., use the gymnasium register function, then use it. or use "lambda: env=env".).

**Comment by MischaPanch** [comment]
Yes, more tutorials and examples will be added soon to help with custom envs. Closing this issue
